service/matching/taskListManager.go

```diff
@@ -385,7 +385,10 @@ func (c *taskListManagerImpl) AddTask(execution *s.WorkflowExecution, taskInfo *
 	if err == nil {
 		c.signalNewTask()
 	}
-	return err
+	if r != nil && err == nil {
```

Comment:
This is incorrect because non-sync match can also go into this. We may need a closure instead of using `r`

Reply:
Can we also add a test to make sure it return Ture when expected?

Reply:
I think this won't happen for non-syncmatch. But if you want to be sure, we can add a field to the response struct and check that field here.

Reply:
Correct me if I am wrong: 
executeWithRetry return its operation's result and err, and it's operation could be either `c.trySyncMatch(taskInfo)` or `c.taskWriter.appendTask(execution, taskInfo, rangeID)`. 
The latter `appendTask` can return `r != nil && err == nil`, which will lead to unwanted `syncMatch` be true. 
Not sure why you think this won't happen.

Reply:
Correct me if I am wrong:
executeWithRetry return its operation's result and err, and it's operation could be either c.trySyncMatch(taskInfo) or c.taskWriter.appendTask(execution, taskInfo, rangeID).
The latter appendTask can return r != nil && err == nil, which will lead to unwanted syncMatch be true.
Not sure why you think this won't happen.

Reply:
Actually Bowei is correct. @longquanzheng , could you add a field to the CreateTasksResponse to indicate if it is sync match or not.

Reply:
Maybe not need a new field in response, just use closure. Inside trySyncMatch if `if (err != nil && err != errAddTasklistThrottled) || r != nil {` add `syncMatch = true`
after `r, err = c.taskWriter.appendTask(execution, taskInfo, rangeID)` add `syncMatch = false`. This may work